683 update access esm how to run (#726)
ACCESS-NRI/access-hive.org.au: docs/models/run-a-model/run-access-esm.md

> {% set model = "ACCESS-ESM1.5" %}

Reviewer: Solved in my recent [commit](https://github.com/ACCESS-NRI/access-hive.org.au/pull/726/commits/76bdfbfa2e885d412914e9b84e218b960ad752af)

## Edit {{ model }} configuration

This section describes how to modify {{ model }} configuration.<br>
The modifications discussed in this section can change the way {{ model }} is run by _payu_, or how its specific [model components] are configured and coupled together.

The `config.yaml` file located in the _control_ directory is the _Master Configuration_ file, which controls the general model configuration. It contains several parts, some of which it is more likely will need modification, and others which are rarely changed without having a deep understanding of how the model is configured.

To find out more about configuration settings for the `config.yaml` file, refer to [how to configure your experiment with payu](https://payu.readthedocs.io/en/latest/config.html).

### Run lengths {: id="runtime"}

One of the most common changes is to adjust the duration of the model run.<br> {{model}} simulations are split into smaller _run lengths_, each with the duration specified by the `runtime` settings in the `config.yaml` file:

The length of an {{model}} run is controlled by the `runtime` settings in the `config.yaml` file:

> ```

Reviewer: Solved in my recent [commit](https://github.com/ACCESS-NRI/access-hive.org.au/pull/726/commits/76bdfbfa2e885d412914e9b84e218b960ad752af)

    ```
    payu run -f -n 7
    ```
    This will submit subsequent jobs for the following years: 1 to 3, 4 to 6, and 7, which is a total of 3 PBS jobs.
!!! tip
    The `walltime` must be set to be long enough that the PBS job can complete. The model usually runs a single year in 90 minutes or less, but the `walltime` for a single model run is set to `2:30:00` out of an abundance of caution to make sure the model has time to run when there are occasional slower runs for unpredicatable reasons. When setting `runspersub > 1` the `walltime` doesn't need to be a simple multiple of `2:30:00` because it is highly unlikely that there will be multiple anomalously slow runs per submit.
### Running for less than one year {: id="shorter-runs"}
When debugging changes to a model, it is common to reduce the run length to minimise resource consumption and return faster feedback on changes. In order to run the model for a single month, the `runtime` can be changed to

> ```

Reviewer: Solved in my recent [commit](https://github.com/ACCESS-NRI/access-hive.org.au/pull/726/commits/76bdfbfa2e885d412914e9b84e218b960ad752af)

    <br><br>
    To run {{ model }}, you need to be a member of a project with allocated _Service Units (SU)_. For more information, check [how to join relevant NCI projects](/getting_started/first_steps#join-relevant-nci-projects).

### Syncing output data

The _laboratory_ directory is typically under the `/scratch` storage on _Gadi_, where [files are regularly deleted once they have been unaccessed for a period of time](https://opus.nci.org.au/pages/viewpage.action?pageId=156434436). For this reason climate model outputs need to be moved to a location with longer term storage.<br>
On _Gadi_, this is typically in a folder under a project code on `/g/data`.  

_Payu_ has built-in support to sync outputs, restarts and a copy of the _control_ directory git history to another location.<br>
This feature is controlled by the following section in the `config.yaml` file: 
```yaml
# Sync options for automatically copying data from ephemeral scratch space to
# longer term storage
sync:
    enable: False # set path below and change to true
    path: null # Set to location on /g/data or a remote server and path (rsync syntax)
```

> To enable syncing, change `enable` to `True`, and set `path` to a location on `/g/data`, where _payu_ will copy output and restart folders. A sensible `path` could be: `/g/data/$PROJECT/$USER/{{model}}/experiment_name/`.

Reviewer 2: ```suggestion
To enable syncing, change `enable` to `True`, and set `path` to a location on `/g/data`, where _payu_ will copy output and restart folders.
```

We don't want users literally putting environment variables in the sync path, so I've suggested we just remove that bit as it got a little confusing trying to suggest something with a bunch of placeholders that needed to be explained and replaced for limited utility.

Author: Good idea, I've just added this in. This section was copied from the OM2 docs – should I add an issue to add the same update there?

#### Runlog

```yaml
runlog: true
```
When running a new configuration, _payu_ automatically commits changes with `git` if `runlog` is set to `true`.

!!! warning
    This should not be changed as it is an essential part of the provenance of an experiment.<br>
    _payu_ updates the manifest files for every run, and relies on `runlog` to save this information in the `git` history, so there is a record of all inputs, restarts, and executables used in an experiment.

#### Userscripts
```yaml
userscripts:
    # Apply land use changes after each run
    run: ./scripts/update_landuse_driver.sh
```

Run scripts or subcommands at various stages of a _payu_ submission. The above example comes from the `release-historical+concentrations` configuration, where the ```update_landuse_driver.sh``` is used to apply historical land use changes at the end of each run.

For more information about specific `userscripts` fields, check the relevant section of [_payu_ Configuration Settings documentation](https://payu.readthedocs.io/en/latest/config.html#postprocessing).

#### Postscripts {: id="postscripts"}

> Postprocessing scripts that need to be run after _payu_ has completed all steps that might alter the output directory can be run as postscripts. These run in separate PBS jobs than the main model simulation. 

Reviewer: Solved in my recent [commit](https://github.com/ACCESS-NRI/access-hive.org.au/pull/726/commits/76bdfbfa2e885d412914e9b84e218b960ad752af)

```yaml
postscript: -v PAYU_CURRENT_OUTPUT_DIR,PROJECT -lstorage=${PBS_NCI_STORAGE} ./scripts/NetCDF-conversion/UM_conversion_job.sh
```

> All {{ model }} configurations include the above NetCDF conversion postscript, which converts the atmosphere model's fields file format output to NetCDF in order to aid analysis and reduce storage requirements. By default, the conversion script will delete the fields files upon successful completion, leaving only the NetCDF output.  This automatic deletion can be disabled by commenting out the `--delete-ff` command line flag from  the conversion job submission script `./scripts/NetCDF-conversion/UM_conversion_job.sh`. That is, changing

Reviewer: Solved in my recent [commit](https://github.com/ACCESS-NRI/access-hive.org.au/pull/726/commits/76bdfbfa2e885d412914e9b84e218b960ad752af)

    <terminal-line>archive path:  /scratch/.../access-esm/archive</terminal-line>
    <terminal-line>Updated metadata. Experiment UUID: 0635396b-678d-45f9-b81e-abd25a2d7bf0</terminal-line>
    <terminal-line>Added archive symlink to /scratch/.../access-esm/archive/preindustrial+concentrations-expt-0635396b</terminal-line>

>     <terminal-line>To change directory to _control_ directory run:</terminal-line>

Reviewer 2: This formatting seems to break in the terminal emulator. The word has disappeared in the preview.

I assume this change was unintentional.

Reviewer: Oops I just did a "replace all" and missed that that was inside a terminal window 😅 
Fixed!

Reviewer 2: I figured.

Now some practical examples:

- **Run 20 years of simulation with resubmission every 5 years**<br>
    To have a _total experiment length_ of 20 years with a 5-year resubmission cycle, leave [`runtime`](#runtime) as the default value of `1 year` and set `runspersub` to `5`. Then, run the configuration with `-n` set to `20`:
    ```
    payu run-f -n 20
    ```
    This will submit subsequent jobs for the following years: 1 to 5, 6 to 10, 11 to 15, and 16 to 20, which is a total of 4 PBS jobs.

- **Run 7 years of simulation with resubmission every 3 years**<br>

>     To have a _total experiment length_ of 7 years with a 3-year resubmission cycle, leave [`runtime`](#runtime) as the default value of `1 year` and set `runspersub` to `3`. Then, run the configuration with `-n` set to `7`:

Reviewer 2: ```suggestion
    To have a _total experiment length_ of 7 years with a 3-year resubmission cycle, leave [`runtime`](#runtime) as the default value of `1 year`, set `runspersub` to `3` and `walltime` to `6:00:00`. Then, run the configuration with `-n` set to `7`:
```

    ```
    payu run -f -n 7
    ```
    This will submit subsequent jobs for the following years: 1 to 3, 4 to 6, and 7, which is a total of 3 PBS jobs.

> 

Reviewer 2: ```suggestion
!!! tip
    The `walltime` must be set to be long enough that the PBS job can complete. The model usually runs a single year in 90 minutes or less, but the `walltime` for a single model run is set to `2:30:00` out of an abundance of caution to make sure the model has time to run when there are occasional slower runs for unpredicatable reasons. When setting `runspersub > 1` the `walltime` doesn't need to be a simple multiple of `2:30:00` because it is highly unlikely that there will be multiple anomalously slow runs per submit.
```

### Running for less than one year {: id="shorter-runs"}
When debugging changes to a model, it is common to reduce the run length to minimise resource consumption and return faster feedback on changes. In order to run the model for a single month, the `runtime` can be changed to

```
    runtime:
        years: 0
        months: 1
        days: 0
```

> With the default configuration settings, the sea ice component of {{ model }} will produce restart files only at the end of each year. If valid restart files are required when running shorter simulations, the sea ice model configuration should be modified so that restart files are produced at monthly frequencies. To do this, change the `dumpfreq = 'y'` setting to `dumpfreq = 'm'` in the `cice_in.nml` configuration file located in the `ice` control directory.

Reviewer 2: ```suggestion
With the default configuration settings, the sea ice component of {{ model }} will produce restart files only at the end of each year. If valid restart files are required when running shorter simulations, the sea ice model configuration should be modified so that restart files are produced at monthly frequencies. To do this, change the `dumpfreq = 'y'` setting to `dumpfreq = 'm'` in the `cice_in.nml` configuration file located in the `ice` subdirectory of the control directory.
```

### Modify PBS resources

If the model has been altered and needs more time to complete, more memory, or needs to be submitted under a different NCI project, you will need to modify the following section in the `config.yaml`:

```yaml
# PBS configuration

# If submitting to a different project to your default, uncomment line below
# and replace PROJECT_CODE with appropriate code. This may require setting shortpath
# project: PROJECT_CODE

# Force payu to always find, and save, files in this scratch project directory
# shortpath: /scratch/PROJECT_CODE

# Note: if laboratory is relative path, it is relative to shortpath/$USER
laboratory: access-esm

jobname: pre-industrial
queue: normal
walltime: 2:30:00
```

These lines can be edited to change the [PBS directives](https://opus.nci.org.au/display/Help/PBS+Directives+Explained) for the [PBS job][PBS job].

> For example, to run {{ model }} under the `ol01` project (COSIMA Working Group), uncomment the line beginning with `# project` by deleting the `#` symbol and replace `PROJECT_CODE` with `ol01`:

Reviewer 2: ```suggestion
By default the model will be submitted to the PBS queue using your default project.  To run {{ model }} using the resources of a specific project, for example the `lg87` project (ESM Working Group), uncomment the line beginning with `# project` by deleting the `#` symbol and replace `PROJECT_CODE` with `lg87`:
```

```yaml

> project: ol01

Reviewer 2: ```suggestion
project: lg87
```

```

!!! warning

>     If a project other than the user's default project is used to run {{ model }} configuration, then the `shortpath` field also needs to be uncommented and the path to the desired `/scratch/PROJECT_CODE` added.<br>
>     Doing this will make sure the same `/scratch` location is used for the _laboratory_, regardless of which project is used to run the experiment.

Reviewer 2: ```suggestion
    If more than one project is used to run an {{ model }} configuration the `shortpath` option also needs to be uncommented and the path to the desired `/scratch/PROJECT_CODE` directory added.<br>
    This ensures the same `/scratch` location is used for the _laboratory_, regardless of which project is used to run the experiment.
```

### Syncing output data

The _laboratory_ directory is typically under the `/scratch` storage on _Gadi_, where [files are regularly deleted once they have been unaccessed for a period of time](https://opus.nci.org.au/pages/viewpage.action?pageId=156434436). For this reason climate model outputs need to be moved to a location with longer term storage.<br>
On _Gadi_, this is typically in a folder under a project code on `/g/data`.  

_Payu_ has built-in support to sync outputs, restarts and a copy of the _control_ directory git history to another location.<br>
This feature is controlled by the following section in the `config.yaml` file: 
```yaml
# Sync options for automatically copying data from ephemeral scratch space to
# longer term storage
sync:
    enable: False # set path below and change to true
    path: null # Set to location on /g/data or a remote server and path (rsync syntax)
```
To enable syncing, change `enable` to `True`, and set `path` to a location on `/g/data`, where _payu_ will copy output and restart folders. A sensible `path` could be: `/g/data/$PROJECT/$USER/{{model}}/experiment_name/`.

!!! Warning

>     The {{model}} default configurations include a postprocessing [postcript](#postscripts) which converts atmospheric outputs to NetCDF format and runs in a separate PBS job. This prevents the final run's output and restart files from being automatically synced. Once the postprocessing is completed, the final outputs and restarts can be manually synced by running `payu sync` from the control directory.

Reviewer 2: ```suggestion
    The {{model}} configurations include a postprocessing [postcript](#postscripts) which converts atmospheric outputs to NetCDF format. This runs in a separate PBS job and this prevents the output and restart files of the most recent run from being automatically synced. When a series of runs completes and the final post-processing is completed run `payu sync` in the control directory to sync the final outputs and restarts.
```

        - /g/data/vk83/configurations/inputs/access-esm1p5/modern/share/coupler/remapping_weights/global.oi_1deg.a_N96/2020.05.19/rmp_cice_to_um1t_CONSERV_FRACNNEI.nc
        - /g/data/vk83/configurations/inputs/access-esm1p5/modern/share/coupler/remapping_weights/global.oi_1deg.a_N96/2020.05.19/rmp_um1u_to_cice_CONSERV_FRACNNEI.nc
        - /g/data/vk83/configurations/inputs/access-esm1p5/modern/share/coupler/remapping_weights/global.oi_1deg.a_N96/2020.05.19/rmp_um1t_to_cice_CONSERV_DESTAREA.nc
        - /g/data/vk83/configurations/inputs/access-esm1p5/modern/share/coupler/remapping_weights/global.oi_1deg.a_N96/2020.05.19/rmp_cice_to_um1u_CONSERV_FRACNNEI.nc
        - /g/data/vk83/configurations/inputs/access-esm1p5/modern/share/coupler/remapping_weights/global.oi_1deg.a_N96/2020.05.19/rmp_um1v_to_cice_CONSERV_FRACNNEI.nc
        - /g/data/vk83/configurations/inputs/access-esm1p5/modern/share/coupler/remapping_weights/global.oi_1deg.a_N96/2020.05.19/rmp_um1t_to_cice_CONSERV_FRACNNEI.nc
        - /g/data/vk83/configurations/inputs/access-esm1p5/modern/share/coupler/remapping_weights/global.oi_1deg.a_N96/2020.05.19/rmp_cice_to_um1v_CONSERV_FRACNNEI.nc

    ```

#### Collate

> Rather than outputting a single diagnostic file over the whole model horizontal grid, the ocean component [MOM](/models/model_components/ocean/#modular-ocean-model-mom) typically generates diagnostic outputs as tiles, each of which spans over a portion of the model horizontal grid.

Reviewer 2: ```suggestion
Rather than outputting a single diagnostic file over the whole model horizontal grid, the ocean component [MOM](/models/model_components/ocean/#modular-ocean-model-mom) typically generates diagnostic outputs as tiles, each of which spans a portion of model grid.
```

The `collate` section in the `config.yaml` file controls the process that combines these smaller files into a single output file.

```yaml
# Collation
collate:
    exe: mppnccombine.spack
    restart: true
    mem: 4GB
    walltime: 1:00:00
    mpi: false
```
Restart files are typically tiled in the same way and will also be combined together if the `restart` field is set to `true`.

#### Runlog

```yaml
runlog: true
```
When running a new configuration, _payu_ automatically commits changes with `git` if `runlog` is set to `true`.

!!! warning
    This should not be changed as it is an essential part of the provenance of an experiment.<br>
    _payu_ updates the manifest files for every run, and relies on `runlog` to save this information in the `git` history, so there is a record of all inputs, restarts, and executables used in an experiment.

#### Userscripts
```yaml
userscripts:
    # Apply land use changes after each run
    run: ./scripts/update_landuse_driver.sh
```

> A dictionary to run scripts or subcommands at various stages of a _payu_ submission. The above example comes from the `release-historical+concentrations` configuration, where the ```update_landuse_driver.sh``` is used to apply historical land use changes at the end of each run.

Reviewer 2: ```suggestion
Run scripts or subcommands at various stages of a _payu_ submission. The above example comes from the `release-historical+concentrations` configuration, where the ```update_landuse_driver.sh``` is used to apply historical land use changes at the end of each run.
```
